2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -86,43 +87,83 @@ A local vault, by default, is already created and named `Default`.
 
 Vault extensions whether they are local or remote are provided as modules.
 
-Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
-`-Credential` or `-AccessToken` parameter for authentication.
-Expectation is that each module will have their own cmdlet and not literally
-named `Get-Secret`.
+Extensions are modules that either:
 
-Extensions register themselves by calling `Register-SecretsVaultExtension`
-passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
-and the module version to `-ModuleVersion`.
-This can be done as part of importing the module or explicitly by the user.
+- Call `Register-SecretsVault` upon loading the module, or
+- Expose a command to the user which calls `Register-SecretsVault` using
+  input from the user
 
-`Register-SecretsVaultExtension` has a `-Default` switch to register the
+Modules that provide a vault extension should have the tag `SecretsVaultExtension`.
+
+Extensions can register themselves by calling `Register-SecretsVault`
+passing the name of the relevant cmdlet to `-Cmdlet` and the module name to `-Module`.
+A `-Name` parameter is used for the user to define a friendly name.
+There is an optional `-Parameters` parameter that takes a hashtable that will
+be splatted to the extension cmdlet to support additional metadata needed
+by the extension (such as authentication).
+
+When using this model, the extension cmdlet would have to match the parameters and
+output of `Get-Secret` to be compatible.
+
+Alternatively, a `-ScriptBlock` parameter can be used instead of `-Cmdlet` and `-Module`
+to allow using existing cmdlets without the need to write a wrapper module:
+
+```powershell
+Register-SecretsVault -Name AzKeyVault {
+  param($Name)
+  Get-AzKeyVaultSecret -VaultName (Get-Secret AzureKeyVaultName) -Name $Name |
+    Select -Expand SecretValue
```

Comment:
The downside to this kind of implementation (as opposed to a C# implementation with standardized implementations) is that there's no consistency as to how things like error handling work.

Can we include a list of standard error messages that providers should use for a few common scenarios such as:

- Access Denied
- Not Found

If providers all implement the exact same error handling for these scenarios, then it will make switching between providers *much* easier, as error handling code won't have to be rewritten.

Reply:
As ideal as standardized error codes are, I do not believe this is something where one could guarantee underlying consistency.  A web-based SecretVault will have a different return code under the covers compared to a C-api based SecretVault.

We _could_ encourage people to use standard -ErrorIDs ( or at least a standard -ErrorCategory) to attempt to address the problem.   Luckily, -ErrorCategory already contains the two we need:  ObjectNotFound and PermissionDenied (or AuthenticationError).

As far as the need/want for a C# api, one can always use PowerShell's C# api to treat these cmdlets as a C# API.  e.g.

    PowerShell.Create().AddScript('Get-Secret MySecret').Invoke<string>();

Thus I believe not only isn't it in scope for this RFC, it shouldn't be for any future RFCs.  I believe to do so opens a nasty pandora's box which would encourage other aspects of PowerShell to provide a C# api as well.  This in turn de-emphasizes the need or want to ever write scripts.

Thus I believe we _should_ specify standard error categories, but _should not_ be in the habit of making special C# wrappers for PowerShell.

Reply:
Cool, the important part of my comment above was the standardization of errors. C# makes it easier to enforce standards, but having a section in the RFC stating that providers should implement the existing error categories/ID's to ensure consistency between different providers is the next best thing.

For example, if one had multiple remote providers and wanted to find a secret but wasn't certain which provider contained it (e.g. during a migration from one provider to another) then they could write a script that iterates through each installed remote provider and puts a "continue" in the catch block if it encounters an exception in the "ObjectNotFound" category. It would suck to have to handle each provider's error in a different way.

Reply:
Agree that standardizing the ErrorRecord will make it easier for scripts to use.  Will add a section on this.